File: backend/app/schemas/smart_savings.py
from pydantic import BaseModel, Field
from typing import List, Optional
from enum import Enum

class TravelMode(str, Enum):
    bus = "bus"
    train = "train"
    flight = "flight"

class SortBy(str, Enum):
    highest_savings = "highest_savings"
    nearest_first = "nearest_first"
    lowest_total = "lowest_total"
    shortest_travel = "shortest_travel"

class SmartSavingsRequest(BaseModel):
    city: str
    procedure: str
    max_nearby_cities: int = Field(default=5, ge=1, le=15)
    max_distance_km: int = Field(default=600, ge=50, le=2000)
    stay_days: int = Field(default=3, ge=1, le=14)
    travel_mode: TravelMode = TravelMode.train
    companions: int = Field(default=1, ge=0, le=5)
    language: str = "en-IN"
    speaker: str = "anushka"
    sort_by: SortBy = SortBy.highest_savings

class TravelCostBreakdown(BaseModel):
    travel_mode: str
    one_way_per_person: float
    round_trip_total: float
    hotel_per_night: float
    hotel_total: float
    total_travel_cost: float
    all_modes: dict  

class NearbyCitySaving(BaseModel):
    rank: int
    city: str
    city_lat: Optional[float] = None
    city_lng: Optional[float] = None
    distance_km: int
    travel_time_hours: float
    travel_time_display: str
    hospital_name: str
    hospital_rating: Optional[float] = None
    procedure_price: float
    travel_cost_breakdown: TravelCostBreakdown
    total_cost: float
    net_savings: float
    savings_percentage: float
    is_worth_it: bool

class NoSavingsCity(BaseModel):
    city: str
    distance_km: int
    cheapest_price: float
    total_cost: float
    reason: str

class UserCityResult(BaseModel):
    city: str
    city_lat: Optional[float] = None
    city_lng: Optional[float] = None
    cheapest_hospital: str
    cheapest_price: float
    most_expensive_hospital: str
    most_expensive_price: float
    all_hospitals: List[dict]

class BestRecommendation(BaseModel):
    city: str
    hospital: str
    total_cost: float
    savings: float
    savings_percentage: float
    message: str

class SmartSavingsResponse(BaseModel):
    success: bool
    user_city: UserCityResult
    nearby_city_savings: List[NearbyCitySaving]
    no_savings_cities: List[NoSavingsCity]
    best_recommendation: Optional[BestRecommendation] = None
    text_response: dict
    voice_response: Optional[dict] = None
    map_data: dict

File: frontend/src/types/smartSavings.ts
export type TravelMode = 'bus' | 'train' | 'flight';
export type SortBy = 'highest_savings' | 'nearest_first' | 'lowest_total' | 'shortest_travel';

export interface SmartSavingsRequest {
  city: string;
  procedure: string;
  max_nearby_cities?: number;
  max_distance_km?: number;
  stay_days?: number;
  travel_mode?: TravelMode;
  companions?: number;
  language?: string;
  speaker?: string;
  sort_by?: SortBy;
}

export interface TravelCostBreakdown {
  travel_mode: string;
  one_way_per_person: number;
  round_trip_total: number;
  hotel_per_night: number;
  hotel_total: number;
  total_travel_cost: number;
  all_modes: Record<string, {
    one_way_per_person: number;
    round_trip_total: number;
    total_with_hotel: number;
  }>;
}

export interface NearbyCitySaving {
  rank: number;
  city: string;
  city_lat: number | null;
  city_lng: number | null;
  distance_km: number;
  travel_time_hours: number;
  travel_time_display: string;
  hospital_name: string;
  hospital_rating: number | null;
  procedure_price: number;
  travel_cost_breakdown: TravelCostBreakdown;
  total_cost: number;
  net_savings: number;
  savings_percentage: number;
  is_worth_it: boolean;
}

export interface NoSavingsCity {
  city: string;
  distance_km: number;
  cheapest_price: number;
  total_cost: number;
  reason: string;
}

export interface UserCityResult {
  city: string;
  city_lat: number | null;
  city_lng: number | null;
  cheapest_hospital: string;
  cheapest_price: number;
  most_expensive_hospital: string;
  most_expensive_price: number;
  all_hospitals: Array<{ hospital_name: string; price: number; rating?: number | null }>;
}

export interface BestRecommendation {
  city: string;
  hospital: string;
  total_cost: number;
  savings: number;
  savings_percentage: number;
  message: string;
}

export interface MapMarker {
  type: 'user_city' | 'savings_city';
  name: string;
  lat: number | null;
  lng: number | null;
  hospital: string;
  price: number;
  total_cost?: number;
  savings?: number;
  savings_pct?: number;
  rank?: number;
  color: string;
}

export interface SmartSavingsResponse {
  success: boolean;
  user_city: UserCityResult;
  nearby_city_savings: NearbyCitySaving[];
  no_savings_cities: NoSavingsCity[];
  best_recommendation: BestRecommendation | null;
  text_response: { english: string; localized: string };
  voice_response: { audio_base64: string; format: string; language: string; sample_rate: number } | null;
  map_data: { center: { lat: number; lng: number }; zoom: number; markers: MapMarker[] };
}
